carbonate_sync.py

```diff
@@ -0,0 +1,318 @@
+    return _run(ssh_cmd)
+
+
+def _filter_metrics(metrics, exclude):
+    """Exclude metrics matching a list of regexp."""
+    for pattern in exclude:
+        regex = re.compile(pattern)
+        metrics = [m for m in metrics if not regex.match(m)]
+    return metrics
+
+
+def _list_metrics(config, local_cluster, local_node, remote_node, exclude,
+                  ssh_options=[]):
+    """List metrics on a remote node."""
+
+    metrics = _ssh(remote_node, ['carbon-list'], ssh_options=ssh_options)
+    metrics = _filter_metrics(metrics, exclude)
+
+    cluster = carbonate_cluster.Cluster(config, local_cluster)
+    metrics = carbonate_sieve.filterMetrics(
+        metrics, local_node, cluster, invert=False)
+    return sorted(metrics)
+
+
+def info(msg=''):
+    logging.info(msg)
+
+
+def _fetch_from_remote(remote_user, remote_host, metrics_fs, staging_dir,
+                       rsync_options=RSYNC_OPTIONS, ssh_options=SSH_OPTIONS):
+    """Fetch a remote list of files using rsync."""
+    remote = '%s@%s:%s/' % (remote_user, remote_host, STORAGE_DIR)
+    with tempfile.NamedTemporaryFile('w', delete=False) as f:
+        for m in metrics_fs:
+            print(m, file=f)
+        f.flush()
+        ssh_cmd = 'ssh ' + ' '.join(ssh_options)
+        cmd = [
+            'rsync',
+            '--rsh', ssh_cmd,
+            '--files-from=%s' % f.name,
+        ]
+        cmd += rsync_options
+        cmd += [
+            remote,
+            staging_dir,
+        ]
+        _run(cmd)
+
+
+def _make_staging_dir(base_dir, node):
+    """Creates the staging directory for this node."""
+    dirname = os.path.join(base_dir, node)
+    try:
+        os.makedirs(dirname)
+    except OSError as exc:
+        # Ignore already existing directories
+        if exc.errno != errno.EEXIST or not os.path.isdir(dirname):
+            raise
+    return dirname
+
+
+def _heal(args_and_metric):
+    """Heal whisper files.
+
+    This method will backfill data present in files in the
+    staging dir if not present in the local files for
+    points between 'start' and 'stop' (unix timestamps).
+    """
+    args, metric = args_and_metric
+    staging_dir, start, end = args
+    carbonate_sync.heal_metric(
+        os.path.join(staging_dir, metric),
+        os.path.join(STORAGE_DIR, metric),
+        # TODO(c.chary): implement that in carbonate
+        # start=start, end=end
+    )
+
+
+def _merge_from_staging(staging_dir, metrics_fs, update_cb, chunksize,
+                        start, end):
+    """Merge files in the staging directory with local files."""
+    workers = multiprocessing.Pool()
+
+    # An iterable of pairs of ((staging_dir, start, end), metric)
+    args = staging_dir, start, end
+    args_and_metrics = itertools.izip(itertools.repeat(args), metrics_fs)
+
+    done = 0
+    update_cb(done)
+
+    if not PARALLEL:
+        for i in args_and_metrics:
+            _heal(i)
+        return
+
```

Comment:
Nop, there is an early return (because it's a very unlikely branch, so I don't want this to affect the normal branch)

PARALLEL = False has been useful multiple time for debugging because it allows to easily put 'print' statements in the libs and display them directly (which isn't that trivial when using multiprocessing).